src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -843,69 +825,47 @@ private void checkout(GitSCM scm, Run<?,?> build, GitClient git, TaskListener li
-             * for membership, paying the performance penalty without the benefit of the accuracy.
-             */
-
-            if (collaboratorNames == null) {
-                listener.getLogger().format("Connecting to %s to obtain list of collaborators for %s/%s%n",
-                        apiUri == null ? GITHUB_URL : apiUri, repoOwner, repository);
+            boolean trusted;
+            PullRequestSCMHead head = (PullRequestSCMHead) revision.getHead();
+            String sourceOwner = head.getSourceOwner();
+            if (repoOwner.equals(sourceOwner)) {
+                // Not a fork, so definitely trusted.
+                trusted = true;
+            } else {
                 StandardCredentials credentials = Connector.lookupScanCredentials(getOwner(), apiUri, scanCredentialsId);
-                // Github client and validation
                 GitHub github = Connector.connect(apiUri, credentials);
+                String fullName = repoOwner + "/" + repository;
+                final GHRepository repo = github.getRepository(fullName);
                 try {
-                    github.checkApiUrlValidity();
-                } catch (HttpException e) {
-                    listener.getLogger().format("It seems %s is unreachable, assuming no trusted collaborators%n",
-                            apiUri == null ? GITHUB_URL : apiUri);
-                    collaboratorNames = Collections.singleton(repoOwner);
-                }
-                if (collaboratorNames == null) {
-                    // Input data validation
-                    if (credentials != null && !github.isCredentialValid()) {
-                        listener.getLogger().format("Invalid scan credentials %s to connect to %s, "
-                                        + "assuming no trusted collaborators%n",
-                                CredentialsNameProvider.name(credentials), apiUri == null ? GITHUB_URL : apiUri);
-                        collaboratorNames = Collections.singleton(repoOwner);
+                    String permission = repo.getPermission(sourceOwner).getPermission();
+                    listener.getLogger().format("Permission check of %s on %s/%s says: %s%n", sourceOwner, repoOwner, repository, permission);
+                    trusted = permission.equals("admin") || permission.equals("write");
+                } catch (FileNotFoundException x) {
+                    listener.getLogger().format("Failed to do a direct permission check of %s on %s/%s; scan token has no access to this repository, or endpoint removed%n", sourceOwner, repoOwner, repository);
+                    trusted = false; // TODO JENKINS-37608: make this configurable
+                } catch (HttpException x) {
+                    if (x.getResponseCode() == 403) {
+                        listener.getLogger().format("Not authorized to do a direct permission check of %s on %s/%s; scan token has no access to this repository%n", sourceOwner, repoOwner, repository);
```

Comment:
> Cache this so you don't retry the request for all PRs

Not sure why caching would be required; this code should only be run during an actual build of the branch project. We could _try_ to cache 403s per repository × `credentialsId`, though this cache could become stale for several reasons—different token, perhaps different permissions for a given token, different repository or team settings…